### PR TITLE
[QC-500] Benchmark module and script

### PR DESCRIPTION
**Pull request by knopers8**
This the first version of the benchmark script for QC task runners. I will come up later with another PR for checkers.
In principle the script can be extended to benchmark any task if other config file template and data producer are given.

I will run the full benchmarks over the weekend, maybe some errors will still come up, but I think it is ready for review like that.

### CODE REVIEW

**Comment by knopers8** [comment]
Needs https://github.com/AliceO2Group/QualityControl/pull/561

File: Modules/Benchmark/include/Benchmark/AlwaysGoodCheck.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+///
+/// \file   BenchmarkCheck.h
```
Comment (Barthelemy) [inline Modules/Benchmark/include/Benchmark/AlwaysGoodCheck.h:12]: name does not match

File: Modules/Benchmark/script/o2-qc-benchmark-tasks.sh [hunk 2]
```diff
@@ -0,0 +1,301 @@
+#!/usr/bin/env bash
+
+#set -e ;# exit on error
+set -u ;# exit when using undeclared variable
+#set -x ;# debugging
+# todo check if bc is installed
+trap kill_benchmark INT
+
+function check_installed() {
+  # very stupid but easy way to check if a package is installed
```
Comment (Barthelemy) [inline Modules/Benchmark/script/o2-qc-benchmark-tasks.sh:10]: I like it :)

**Comment by Barthelemy** [comment]
Good thanks !

**Review by Barthelemy — APPROVED** [review]

**Comment by ktf** [comment]
@TimoWilken can you check what is wrong with the mac builds?

**Comment by knopers8** [comment]
I've just noticed a bug in DPL which makes the benchmarks fail in some configuration, I will make it a WIP for now.

**Comment by ktf** [comment]
Can you elaborate on the bug, please?

**Comment by knopers8** [comment]
Still experimenting to understand when it happens exactly, but so far it seems to appear when I have at least two producers with the same origin and description, but different subspec. The receiver subscribes to these and has a wildcard on subspec.
When the message rate is high enough (>1000/s), I see a flood of these on the receiver side:
```
[23956:QC-TASK-RUNNER-BenchmarkTask]: [12:07:46][WARN] Incoming data is already obsolete, not relaying.
[23956:QC-TASK-RUNNER-BenchmarkTask]: [12:07:46][WARN] Incoming data is already obsolete, not relaying.
[23956:QC-TASK-RUNNER-BenchmarkTask]: [12:07:46][WARN] Incoming data is already obsolete, not relaying.

```
Eventually there is a segmentation violation.

Was the related part of code modified recently?

**Comment by ktf** [comment]
The only change is that now they are multiplexed on the same multipart message. This was done a while ago, though. What is the crash about?

**Comment by knopers8** [comment]
@ktf The crash seems to be related to the code which reports errors:
```
[23594:QC-TASK-RUNNER-BenchmarkTask]: Thread 1 (Thread 0x7f80c798d000 (LWP 23594)):
[23594:QC-TASK-RUNNER-BenchmarkTask]: #0  0x00007f80b78fc469 in waitpid () from /lib64/libc.so.6
[23594:QC-TASK-RUNNER-BenchmarkTask]: #1  0x00007f80b7879f12 in do_system () from /lib64/libc.so.6
[23594:QC-TASK-RUNNER-BenchmarkTask]: #2  0x00007f80b787a2c1 in system () from /lib64/libc.so.6
[23594:QC-TASK-RUNNER-BenchmarkTask]: #3  0x00007f80b87695b0 in TUnixSystem::Exec (shellcmd=<optimized out>, this=0x1036cf0) at /home/pkonopka/alice/sw/SOURCES/ROOT/v6-20-08-alice1/v6-20-08-alice1/core/un
ix/src/TUnixSystem.cxx:2107
[23594:QC-TASK-RUNNER-BenchmarkTask]: #4  TUnixSystem::StackTrace (this=0x1036cf0) at /home/pkonopka/alice/sw/SOURCES/ROOT/v6-20-08-alice1/v6-20-08-alice1/core/unix/src/TUnixSystem.cxx:2397
[23594:QC-TASK-RUNNER-BenchmarkTask]: #5  0x00007f80b876bc44 in TUnixSystem::DispatchSignals (this=0x1036cf0, sig=kSigSegmentationViolation) at /home/pkonopka/alice/sw/SOURCES/ROOT/v6-20-08-alice1/v6-20-0
8-alice1/core/unix/src/TUnixSystem.cxx:3628
[23594:QC-TASK-RUNNER-BenchmarkTask]: #6  <signal handler called>
[23594:QC-TASK-RUNNER-BenchmarkTask]: #7  0x00007f80c3f64bf2 in o2::framework::DataProcessingDevice::<lambda(char const*)>::operator()(const char *) const (__closure=__closure
[23594:QC-TASK-RUNNER-BenchmarkTask]: entry=0x7ffc7cb029b0, message=<optimized out>) at /home/pkonopka/alice/sw/SOURCES/O2/0_O2_DATAFLOW/0/Framework/Core/src/DataProcessingDevice.cxx:730
[23594:QC-TASK-RUNNER-BenchmarkTask]: #8  0x00007f80c3f6b6ca in o2::framework::DataProcessingDevice::<lambda(const std::vector<o2::framework::DataProcessingDevice::handleData(o2::framework::DataProcessorC
ontext&, FairMQParts&, o2::framework::InputChannelInfo&)::InputType, std::allocator<o2::framework::DataProcessingDevice::handleData(o2::framework::DataProcessorContext&, FairMQParts&, o2::framework::Input
ChannelInfo&)::InputType> >&)>::operator() (types=..., __closure=<optimized out>) at /home/pkonopka/alice/sw/SOURCES/O2/0_O2_DATAFLOW/0/Framework/Core/src/DataProcessingDevice.cxx:745
[23594:QC-TASK-RUNNER-BenchmarkTask]: #9  o2::framework::DataProcessingDevice::handleData (context=..., parts=..., info=...) at /home/pkonopka/alice/sw/SOURCES/O2/0_O2_DATAFLOW/0/Framework/Core/src/DataPr
ocessingDevice.cxx:772
[23594:QC-TASK-RUNNER-BenchmarkTask]: #10 0x00007f80c3f6b9ce in o2::framework::DataProcessingDevice::doPrepare (context=...) at /home/pkonopka/alice/sw/SOURCES/O2/0_O2_DATAFLOW/0/Framework/Core/src/DataProcessingDevice.cxx:579
[23594:QC-TASK-RUNNER-BenchmarkTask]: #11 0x00007f80c3f6d715 in o2::framework::run_callback (handle=<optimized out>) at /home/pkonopka/alice/sw/SOURCES/O2/0_O2_DATAFLOW/0/Framework/Core/src/DataProcessing
Device.cxx:136
[23594:QC-TASK-RUNNER-BenchmarkTask]: #12 0x00007f80c3f6d7e9 in o2::framework::DataProcessingDevice::ConditionalRun (this=0x42fb440) at /home/pkonopka/alice/sw/SOURCES/O2/0_O2_DATAFLOW/0/Framework/Core/sr
c/DataProcessingDevice.cxx:515
```
DataProcessingDevice.cxx:730 is:
```
  auto reportError = [& registry = *context.registry, &context](const char* message) {
    context.errorCount++;
    registry.get<Monitoring>().send(Metric{*context.errorCount, "errors"}.addTag(Key::Subsystem, Value::DPL));
  };
```
So I guess it is only a consequence of the large amount of "obsolete" errors. I have assigned to you a jira issue with more information, but I can't point you to it now, because I've lost access to jira by accident.

**Comment by knopers8** [comment]
I have found a workaround for the problem in DPL, I will write about it in the jira issue once i have jira access back (yes, again).
I've added a check runner benchmark and some foundation for PP benchmark (which I will have to finish later, this is not urgent).

**Comment by ktf** [comment]
This https://github.com/AliceO2Group/AliceO2/pull/5371 should reduce the amount of metrics being sent on error.

**Pull request merged**